src/cargo/ops/cargo_compile.rs

```diff
@@ -944,6 +945,48 @@ fn generate_targets<'a>(
     Ok(units.into_iter().collect())
 }
 
+fn warn_on_missing_features(resolve: &Resolve, required_features: &[String], summary: &Summary) {
+    for feature in required_features {
+        match FeatureValue::new(feature.into(), summary) {
+            // No need to do anything here, since the feature must exist to be parsed as such
+            FeatureValue::Feature(_) => {}
+            // Possibly mislabeled feature that was not found
+            FeatureValue::Crate(krate) => {
+                if summary
+                    .dependencies()
+                    .iter()
+                    .map(|d| d.name_in_toml())
+                    .find(|name| name == &krate)
+                    .is_none()
+                {
+                    eprintln!("warning: feature {} not found", krate)
+                }
+            }
+            // Handling of dependent_crate/dependent_crate_feature syntax
+            FeatureValue::CrateFeature(krate, feature) => {
+                match resolve.iter().find(|pkg| pkg.name() == krate) {
```

Comment:
I think this should instead get the dependencies from the resolver instead of using `find`.  There can be multiple packages that have the same name (such as different versions), and the value `krate` here is the *renamed* name ("name_in_toml"), so it needs to look it up that way. It will also need to check both features and optional dependencies. I think it would be something like this:

```rust
match resolve
    .deps(summary.package_id())
    .find(|(_dep_id, deps)| deps.iter().any(|dep| dep.name_in_toml() == krate))
{
    Some((dep_id, _deps)) => {
        let dep_summary = resolve.summary(dep_id);
        if !dep_summary
            .features()
            .contains_key(&feature)
            && !dep_summary.dependencies().iter().any(|dep| dep.name_in_toml() == feature && dep.is_optional())
        {
            eprintln!("warning required-feature \"{}\" is not present in package `{}` \
                (specified in package `{}` for target …yadda yadda…)",
                dep_id, feature, summary.package_id());
        }
    }
    None => eprintln!("warning dependency \"{}\" specified in required-feature \"{}/{}\" does not exist \
        (specified in package `{}` for target  …yadda yadda…)",
        krate, krate, feature, summary.package_id()),
}
```

I'm not sure about the wording of the warnings, that can maybe be tweaked.  Also, I tossed that together quickly, but can probably be cleaned up a little.